### PR TITLE
[SparseAutoModelForCausalLM Deprecation] Feature change

### PR DESCRIPTION
**Pull request by horheynm**
SUMMARY:
Deprecate SparseAutoModelForCausalLM and use AutoModelForCausalLM

SparseAutoModelForCausalLM was responsible for adding custom load and save logic.
Load logic is no longer needed - taken care of in HF transfomers - HFQuantizer
Save logic is kept, wrapping save_pretrained.
Two saving cases -> fsdp model and non-fsdp model. 

Key changes:
* models' save_pretrained in wrapped if and only if the model uses the optimization pipeline (train, oneshot, etc). If not, then no changes so vanilla save_pretained will be used.
* if model is passed as a string, then model will be automatically saved in oneshot, elif passed as a pretrained_model instance, then users can choose to save it manually using save_pretrained
* SparseAutoModelForCausalLM and SparseAutoModel are now deprecated - will trigger a warning when used. Will be removed in the appropriate feature release

TEST PLAN:
"please outline how the changes were tested"

### CODE REVIEW

File: src/llmcompressor/transformers/finetune/README.md
Comment (kylesayrs) [inline src/llmcompressor/transformers/finetune/README.md:64]: It occurs to me that removing saving within the oneshot function also removes user's ability to pass a string name into the model argument, since there is no way to save the compressed model if only a string is passed.

This is critical functionality we want to support, so we should definitely continue to allow users to save via oneshot, and we should also guarantee check arguments to guarantee that if a string is passed for the model argument, then output_dir is not empty

Comment (horheynm) [reply src/llmcompressor/transformers/finetune/README.md:64]: ok

**Review by dsikka — COMMENTED** [review]
Is the feature branch opened up in a PR so that testing can run?

File: tests/llmcompressor/transformers/sparsification/test_compress_tensor_utils.py
Comment (dsikka) [inline tests/llmcompressor/transformers/sparsification/test_compress_tensor_utils.py:144]: why are we commenting out test cases?

File: tests/llmcompressor/transformers/sparsification/test_compress_tensor_utils.py
Comment (dsikka) [inline tests/llmcompressor/transformers/sparsification/test_compress_tensor_utils.py:195]: Why are we removing valid test cases?

File: src/llmcompressor/transformers/finetune/text_generation.py
Comment (dsikka) [inline src/llmcompressor/transformers/finetune/text_generation.py:213]: Is there a need a to keep SparseAutoModel if we're swapping it to just use AutoModelForCausalLM?

**Review by dsikka — CHANGES_REQUESTED** [review]
Couple of key points:
1. Why are we changing/commenting valid test cases?
2. From our offline discussion, keeping output_dir is fine, we just want to only be saving once, which it seems like we're doing now with these changes (i.e if an output_dir is defined). We should still provide a string for the model/not require just a model object. 
3. I still need to pull these chantes to test locally.

File: tests/llmcompressor/transformers/sparsification/test_compress_tensor_utils.py
Comment (horheynm) [reply tests/llmcompressor/transformers/sparsification/test_compress_tensor_utils.py:195]: This is the change to AutoModelForCuasalLM

File: src/llmcompressor/transformers/finetune/text_generation.py
Comment (horheynm) [reply src/llmcompressor/transformers/finetune/text_generation.py:215]: Its just for backwards compatibility, not needed for functionality.

**Comment by horheynm** [comment]
@dsikka 
This pr will be merged to a feature branch where tests will run. Tests runner there can validate.

File: tests/llmcompressor/transformers/sparsification/test_compress_tensor_utils.py
Comment (dsikka) [inline tests/llmcompressor/transformers/sparsification/test_compress_tensor_utils.py:206]: why are we changing this

File: tests/llmcompressor/transformers/sparsification/test_compress_tensor_utils.py
Comment (dsikka) [inline tests/llmcompressor/transformers/sparsification/test_compress_tensor_utils.py:186]: this seems like we're now changing what the test was testing before?

File: tests/llmcompressor/transformers/sparsification/test_compress_tensor_utils.py
Comment (dsikka) [inline tests/llmcompressor/transformers/sparsification/test_compress_tensor_utils.py:172]: you should still be able to pass in a string

File: tests/llmcompressor/transformers/obcq/test_mask_structure_preservation.py
Comment (dsikka) [inline tests/llmcompressor/transformers/obcq/test_mask_structure_preservation.py:62]: you should still be able to pass in a string

File: tests/llmcompressor/transformers/gptq/test_oneshot.py
Comment (dsikka) [inline tests/llmcompressor/transformers/gptq/test_oneshot.py:79]: same here

**Review by dsikka — COMMENTED** [review]
From our offline discussion, keeping output_dir is fine, we just want to only be saving once, which it seems like we're partly doing now with these changes (i.e only save as part of the one_shot call if an output_dir is defined). We should still be able to provide a string for the model/not require just a model object

File: tests/llmcompressor/transformers/sparsification/test_compress_tensor_utils.py
Comment (horheynm) [reply tests/llmcompressor/transformers/sparsification/test_compress_tensor_utils.py:172]: You can but the main pathway to run compression and save is to use model.save_pretrained() outside the oneshot. Hence passing in model

File: tests/llmcompressor/transformers/sparsification/test_compress_tensor_utils.py
Comment (horheynm) [reply tests/llmcompressor/transformers/sparsification/test_compress_tensor_utils.py:186]: Previously oneshot pathway does the save, so this is to replicate that behavior

File: tests/llmcompressor/transformers/obcq/test_mask_structure_preservation.py
Comment (horheynm) [reply tests/llmcompressor/transformers/obcq/test_mask_structure_preservation.py:62]: ok

File: tests/llmcompressor/transformers/sparsification/test_compress_tensor_utils.py
Comment (horheynm) [reply tests/llmcompressor/transformers/sparsification/test_compress_tensor_utils.py:144]: reverted

File: tests/llmcompressor/transformers/obcq/test_consecutive_runs.py
Comment (kylesayrs) [inline tests/llmcompressor/transformers/obcq/test_consecutive_runs.py:52]: Please remove
```suggestion
```

**Review by kylesayrs — CHANGES_REQUESTED** [review]

File: tests/llmcompressor/transformers/compression/test_run_compressed.py
Comment (dsikka) [inline tests/llmcompressor/transformers/compression/test_run_compressed.py:24]: run_compressed should already be True, correct?

File: tests/llmcompressor/transformers/compression/test_quantization.py
Comment (dsikka) [inline tests/llmcompressor/transformers/compression/test_quantization.py:104]: I think you're changing what the test is actually supposed to be testing here. By saving the model from the oneshot and not loading it from disk, we're not able to test the actual reloading to verify that it was done correctly.

File: tests/llmcompressor/transformers/sparsification/test_compress_tensor_utils.py
Comment (dsikka) [inline tests/llmcompressor/transformers/sparsification/test_compress_tensor_utils.py:314]: Why do we now need this patch?

File: src/llmcompressor/transformers/finetune/README.md
Comment (dsikka) [inline src/llmcompressor/transformers/finetune/README.md:119]: Considering the example is using a string, we should keep the output dir since we're not showing the saving step in the readme

**Comment by dsikka** [comment]
Overall testing locally looks good. I think this should be good to land once the last few comments are addressed.

**Review by dsikka — CHANGES_REQUESTED** [review]
It seems like we're not saving the output for quantization or sparsity from the apply() call when running:
`llm-compressor/examples/quantization_2of4_sparse_w4a16/llama7b_sparse_w4a16.py`. This is currently saved to the output dir on main

File: tests/llmcompressor/transformers/compression/test_quantization.py
Comment (horheynm) [reply tests/llmcompressor/transformers/compression/test_quantization.py:104]: Theres no way to load a quantized model using HFQuantizer at this time, which is I am gonna be working on next.
Added a TODO as comment

File: tests/llmcompressor/transformers/sparsification/test_compress_tensor_utils.py
Comment (horheynm) [reply tests/llmcompressor/transformers/sparsification/test_compress_tensor_utils.py:314]: SparseAutoModelForCausalLM on load ran this code before

File: tests/llmcompressor/transformers/compression/test_quantization.py
Comment (dsikka) [reply tests/llmcompressor/transformers/compression/test_quantization.py:110]: You can reload using the ModelCompressor and put a TODO to swap in to from_pretrained.

File: src/llmcompressor/transformers/sparsification/compressed_tensors_utils.py
Comment (kylesayrs) [inline src/llmcompressor/transformers/sparsification/compressed_tensors_utils.py:146]: This needs to go at the start of oneshot, since we need to untie the tensors before compression starts

**Review by kylesayrs — CHANGES_REQUESTED** [review]

**Review by kylesayrs — APPROVED** [review]

File: src/llmcompressor/transformers/finetune/README.md
Comment (rahul-tuli) [inline src/llmcompressor/transformers/finetune/README.md:103]:
```suggestion
model = "Xenova/llama2.c-stories15M"
```

File: src/llmcompressor/transformers/finetune/README.md
Comment (rahul-tuli) [inline src/llmcompressor/transformers/finetune/README.md:162]: Why not keep this here if it's still supported?

File: src/llmcompressor/transformers/finetune/text_generation.py
Comment (rahul-tuli) [inline src/llmcompressor/transformers/finetune/text_generation.py:396]: Could you confirm if this saves the recipe as well?

File: src/llmcompressor/transformers/sparsification/compressed_tensors_utils.py
Comment (rahul-tuli) [inline src/llmcompressor/transformers/sparsification/compressed_tensors_utils.py:102]: Misleading docstring!

File: src/llmcompressor/transformers/sparsification/compressed_tensors_utils.py
Comment (rahul-tuli) [inline src/llmcompressor/transformers/sparsification/compressed_tensors_utils.py:247]: Missing docstring

File: tests/llmcompressor/transformers/compression/test_quantization.py
Comment (rahul-tuli) [inline tests/llmcompressor/transformers/compression/test_quantization.py:104]: Is this TODO still relevant?

**Review by rahul-tuli — COMMENTED** [review]
LGTM pending comments!

File: tests/llmcompressor/transformers/compression/test_quantization.py
Comment (horheynm) [reply tests/llmcompressor/transformers/compression/test_quantization.py:104]: Yes

File: src/llmcompressor/transformers/sparsification/sparse_model.py
Comment (horheynm) [reply src/llmcompressor/transformers/sparsification/sparse_model.py:14]: No need

File: src/llmcompressor/transformers/finetune/README.md
Comment (horheynm) [reply src/llmcompressor/transformers/finetune/README.md:162]: yes you are right

File: src/llmcompressor/transformers/finetune/text_generation.py
Comment (horheynm) [reply src/llmcompressor/transformers/finetune/text_generation.py:396]: yes

File: src/llmcompressor/transformers/sparsification/compressed_tensors_utils.py
Comment (horheynm) [reply src/llmcompressor/transformers/sparsification/compressed_tensors_utils.py:102]: This is fine, it does wrap the save_pretrained

File: tests/llmcompressor/transformers/sparsification/test_compress_tensor_utils.py
Comment (horheynm) [reply tests/llmcompressor/transformers/sparsification/test_compress_tensor_utils.py:145]: This test is outdated. Three input args vs two

File: tests/llmcompressor/transformers/compression/test_quantization.py
Comment (dsikka) [inline tests/llmcompressor/transformers/compression/test_quantization.py:108]: should only use cuda if available

File: src/llmcompressor/transformers/finetune/text_generation.py
Comment (horheynm) [reply src/llmcompressor/transformers/finetune/text_generation.py:396]: yes its saved

File: src/llmcompressor/transformers/sparsification/compressed_tensors_utils.py
Comment (rahul-tuli) [reply src/llmcompressor/transformers/sparsification/compressed_tensors_utils.py:102]: The typing says model is an nn.Module, but the docstring says we override a PretrainedModel's save_pretrained?

**Review by dsikka — COMMENTED** [review]
Example cases all look good. Will look through code once more before approving.

File: tests/llmcompressor/transformers/finetune/test_oneshot_and_finetune_with_tokenizer.py
Comment (dsikka) [inline tests/llmcompressor/transformers/finetune/test_oneshot_and_finetune_with_tokenizer.py:41]: Test uses SparseAutoModelForCausalLM still? Line 41/42

File: src/llmcompressor/transformers/finetune/README.md
Comment (dsikka) [reply src/llmcompressor/transformers/finetune/README.md:64]: add back output dir to the example

**Review by dsikka — COMMENTED** [review]
Two open comments. Otherwise, LGTM

**Comment by kylesayrs** [comment]
tests/llmcompressor/transformers/compression/test_run_compressed.py
ests/llmcompressor/transformers/finetune/test_finetune_no_recipe_custom_dataset.py

These tests are newly failing

File: tests/llmcompressor/transformers/finetune/test_oneshot_and_finetune_with_tokenizer.py
Comment (horheynm) [reply tests/llmcompressor/transformers/finetune/test_oneshot_and_finetune_with_tokenizer.py:41]: good catch, merge overwrite

**Comment by horheynm** [comment]
> tests/llmcompressor/transformers/compression/test_run_compressed.py ests/llmcompressor/transformers/finetune/test_finetune_no_recipe_custom_dataset.py
> 
> These tests are newly failing

Good catch, thank you

**Comment by dsikka** [comment]
also seems like sparsification test now have an issue with loading some of the weights

File: tests/llmcompressor/transformers/compression/test_run_compressed.py
Comment (kylesayrs) [inline tests/llmcompressor/transformers/compression/test_run_compressed.py:24]: I would just skip this test for now, then reenable when run_compressed is fully supported

File: src/llmcompressor/transformers/sparsification/compressed_tensors_utils.py
Comment (kylesayrs) [inline src/llmcompressor/transformers/sparsification/compressed_tensors_utils.py:78]: Note that we should investigate why we don't need model_ref anymore

File: src/llmcompressor/transformers/sparsification/compressed_tensors_utils.py
Comment (kylesayrs) [inline src/llmcompressor/transformers/sparsification/compressed_tensors_utils.py:248]: Cleaner! :)

**Review by kylesayrs — APPROVED** [review]
Looks good to me, I made a few comments about places to investigate after this lands

File: tests/llmcompressor/transformers/sparsification/test_compress_tensor_utils.py
Comment (dsikka) [inline tests/llmcompressor/transformers/sparsification/test_compress_tensor_utils.py:66]: I dont think we want to turn off warnings as they're helpful in understanding what's going on.

Comment (horheynm) [reply tests/llmcompressor/transformers/sparsification/test_compress_tensor_utils.py:66]: https://github.com/vllm-project/llm-compressor/blob/93832a662322e3f6bbc246ba4e7d960faeb9510b/src/llmcompressor/transformers/sparsification/sparse_model.py#L90-L123

**Review by dsikka — COMMENTED** [review]
LGTM. Just last two comments/questions

File: src/llmcompressor/transformers/finetune/README.md
Comment (horheynm) [reply src/llmcompressor/transformers/finetune/README.md:122]: its a duplicate

File: src/llmcompressor/transformers/finetune/runner.py
Comment (horheynm) [reply src/llmcompressor/transformers/finetune/runner.py:264]: It is added already, save_pretrained is wrapped based on if its fsdp or not

**Review by dsikka — APPROVED** [review]
Nice job on getting all the flows working.

**Review by rahul-tuli — APPROVED** [review]

**Pull request merged**
